File: crates/dcl/src/js/modules/SystemApi.js
// system api module

const { 
    op_check_for_update, op_motd, 
    op_get_current_login, op_get_previous_login, op_login_previous, op_login_new_code, op_login_new_success, op_login_cancel, op_login_guest, op_logout,
    op_settings, op_set_setting,
} = Deno.core.ops;

// (description: option<string>, url: option<string>)
module.exports.checkForUpdate = async function() {
    const [description, url] = await op_check_for_update();
    return {
        description,
        url
    }
}

// (message: string)
module.exports.messageOfTheDay = async function() {
    return {
        message: await op_motd()
    }
}

// (currentLogin: option<string>)
module.exports.getCurrentLogin = function() {
    return {
        current_login: op_get_current_login()
    }
}

// (userId: option<string>)
module.exports.getPreviousLogin = async function() {
    return {
        userId: await op_get_previous_login()
    }
}

// (success: bool, error: option<string>)
module.exports.loginPrevious = async function() {
    return await op_login_previous()
}

// (code: Promise<option<string>>, success: Promise<(success: bool, error: option<string>)>)
module.exports.loginNew = function() {
    return {
        code: op_login_new_code(),
        success: op_login_new_success(),
    }
}

// nothing
module.exports.loginGuest = function() {
    op_login_guest()
}

// nothing
module.exports.loginCancel = function() {
    op_login_cancel()
}

// nothing
module.exports.logout = function() {
    op_logout()
}

// array of {
//   name: string, 
//   category: string, 
//   description: string,
//   minValue: number, 
//   maxValue: number, 
//   namedVariants: [(variantName: string, variantDescription: string)]
//   value: number,
//   default: number,
// }
module.exports.getSettings = async function() {
    return await op_settings();
}

module.exports.setSetting = async function(name, value) {
    await op_set_setting(name, value);
}

module.exports.kernelFetch = async function (body) { 
    const headers = await Deno.core.ops.op_kernel_fetch_headers(body.url, body.init?.method, body.meta);

    if (!body.init) {
        body.init = { headers: {} };
    }

    if (!body.init.hasOwnProperty("headers")) {
        body.init.headers = {};
    }

    for (var i=0; i< headers.length; i++) {
        body.init.headers[headers[i][0]] = headers[i][1];
    }

    let response = await fetch(body.url, body.init);
    let text = await response.text();

    return {
        ok: response.ok,
        status: response.status,
        statusText: response.statusText,
        headers: response.headers,
        body: text,
    };
}

// avatar { 
//   base?: PBAvatarBase, 
//   equip?: PBAvatarEquippedData,
//   hasClaimedName?: bool,
//   profileExtras?: {field: value}
// }
// => deployed version
module.exports.setAvatar = async function(avatar) {
    return await Deno.core.ops.op_set_avatar(avatar.base, avatar.equip, avatar.hasClaimedName, avatar.profileExtras)
}

module.exports.getProfileExtras = async function() {
    return await Deno.core.ops.op_get_profile_extras();
}

// get the next key/button pressed by the user, identified as a string
// -> string
module.exports.getNativeInput = async function() {
    return await Deno.core.ops.op_native_input()
}

// get current key bindings
// -> {
//   bindings: (string, string[])[]
// }
module.exports.getInputBindings = async function() {
    return await Deno.core.ops.op_get_bindings()
}

// set current key bindings
// arg: {
//   bindings: (string, string[])[]
// }
module.exports.setInputBindings = async function(bindings) {
    await Deno.core.ops.op_set_bindings(bindings)
}


/// reload 
// hash: string | undefined
// if hash is provided, that specific scene is reloaded
// if unspecified, all scenes are reloaded
module.exports.reload = async function(hash) {
    await Deno.core.ops.op_console_command("reload", hash !== undefined ? [hash] : [])
}

// show_ui
// {
//   hash: string | undefined,
//   show: bool | undefined
// }
//
// if hash is undefined, all scenes are modified
// if show is undefined, acts as a toggle
// returns: visible bool 
// note: doesn't hide/show the system ui scene
module.exports.showUi = async function(args) {
  let argsArray = [args?.hash ?? "all"]
  if (args?.show !== undefined) {
    argsArray.push(args.show ? "true" : "false")
  }

  const reply = await Deno.core.ops.op_console_command("show_ui", argsArray)
  const value = reply.split(":").pop()?.trim().toLowerCase();
  return value === "true";  
}

// [{
//     hash: string,
//     base_url?: string,
//     title: string,
//     parcels: [Vector2],
//     isPortable: bool,
//     isBroken: bool,
//     isBlocked: bool,
//     isSuper: bool,
//     sdkVersion: string,
// }]
module.exports.liveSceneInfo = async function() {
    return await Deno.core.ops.op_live_scene_info()
}

// { 
//   realm: string, // about_url (e.g. `realm-provider.decentraland.org/main`)
//   parcel: Vector2,
// }
module.exports.getHomeScene = async function() {
    return await Deno.core.ops.op_get_home_scene()
}

// { 
//   realm: string, // about_url (e.g. `realm-provider.decentraland.org/main`)
//   parcel: Vector2,
// }
module.exports.setHomeScene = async function(args) {
    await Deno.core.ops.op_set_home_scene(args.realm, args.parcel)
}

// get system actions as a stream
// type SystemAction = {
//   action: string,
//   pressed: boolean,
// }
module.exports.getSystemActionStream = async function() {
  const rid = await Deno.core.ops.op_get_system_action_stream();

  async function* streamGenerator() {
    while (true) {
      const next = await Deno.core.ops.op_read_system_action_stream(rid);
      if (next === null) break;
      yield next;
    }
  }

  return streamGenerator();
}

// get chat messages as a stream
// type ChatMessage = {
//   senderAddress: string,
//   message: string,
//   channel: string,
// }
module.exports.getChatStream = async function() {
  const rid = await Deno.core.ops.op_get_chat_stream();

  async function* streamGenerator() {
    while (true) {
      const next = await Deno.core.ops.op_read_chat_stream(rid);
      if (next === null) break;
      yield next;
    }
  }

  return streamGenerator();
}

// send a chat message
// { 
//   message: string,
//   channel?: string
// }
module.exports.sendChat = async function(message, channel) {
    Deno.core.ops.op_send_chat(message, channel ?? "Nearby")
}

module.exports.quit = function() {
    Deno.core.ops.op_quit();
}

module.exports.getPermissionRequestStream = async function() {
  const rid = await Deno.core.ops.op_get_permission_request_stream();

  async function* streamGenerator() {
    while (true) {
      const next = await Deno.core.ops.op_read_permission_request_stream(rid);
      if (next === null) break;
      yield next;
    }
  }

  return streamGenerator();
}

module.exports.getPermissionUsedStream = async function() {
  const rid = await Deno.core.ops.op_get_permission_used_stream();

  async function* streamGenerator() {
    while (true) {
      const next = await Deno.core.ops.op_read_permission_used_stream(rid);
      if (next === null) break;
      yield next;
    }
  }

  return streamGenerator();
}

module.exports.setSinglePermission = function(body) {
    Deno.core.ops.op_set_single_permission(body.id, body.allow);
}

module.exports.setPermanentPermission = function(body) {
    Deno.core.ops.op_set_permanent_permission(body.level, body.value, body.ty, body.allow)
}

module.exports.getPermanentPermissions = async function(body) {
    return await Deno.core.ops.op_get_permanent_permissions(body.level, body.value);
}

module.exports.getPermissionTypes = function() {
    return Deno.core.ops.getPermissionTypes();
}

module.exports.setInteractableArea = function(body) {
    Deno.core.ops.op_set_interactable_area(body.left, body.top, body.right, body.bottom);
}

module.exports.getMicState = function() {
    return Deno.core.ops.op_get_mic_state();
}

module.exports.setMicEnabled = function(enabled) {
    Deno.core.ops.op_set_mic_enabled(enabled);
}

// get voice stream / mic activations as a stream
// type MicActivation = {
//   senderAddress: string,
//   active: bool,
// }
module.exports.getVoiceStream = async function() {
  const rid = await Deno.core.ops.op_get_voice_stream();

  async function* streamGenerator() {
    while (true) {
      const next = await Deno.core.ops.op_read_voice_stream(rid);
      if (next === null) break;
      yield next;
    }
  }

  return streamGenerator();
}

// get hover events as a stream
// HoverTargetType: 0 = World, 1 = Ui, 2 = Avatar
// PointerEventType: 0 = PET_UP, 1 = PET_DOWN, 2 = PET_HOVER_ENTER, 3 = PET_HOVER_LEAVE, 4 = PET_DRAG_LOCKED, 5 = PET_DRAG, 6 = PET_DRAG_END
// type HoverEventInfo = {
//   inputAction: number,   // InputAction (0-13)
//   hoverText: string,
//   hideFeedback: boolean,
//   showHighlight: boolean,
//   maxDistance: number,
// }
// type HoverAction = {
//   eventType: number,     // PointerEventType (0-6)
//   eventInfo: HoverEventInfo,
// }
// type HoverEvent = {
//   entered: boolean,
//   targetType: HoverTargetType,
//   distance: number,
//   actions: HoverAction[],
//   outsideScene: boolean,  // true if player is outside the scene containing the entity (always false for avatars)
// }
module.exports.getHoverStream = async function() {
  const rid = await Deno.core.ops.op_get_hover_stream();

  async function* streamGenerator() {
    while (true) {
      const next = await Deno.core.ops.op_read_hover_stream(rid);
      if (next === null) break;
      yield next;
    }
  }

  return streamGenerator();
}

File: crates/dcl_deno/src/js/op_wrappers/system_api.rs
use common::{
    inputs::SystemActionEvent,
    structs::{PermissionType, PermissionUsed, PermissionValue},
};
use dcl::js::system_api::{JsBindingsData, PermissionTypeDetail};
use dcl_component::proto_components::{
    common::Vector2,
    sdk::components::{PbAvatarBase, PbAvatarEquippedData},
};
use deno_core::{anyhow, error::AnyError, op2, OpDecl, OpState};
use std::{cell::RefCell, rc::Rc};
use system_bridge::{
    settings::SettingInfo, ChatMessage, HomeScene, LiveSceneInfo, PermanentPermissionItem,
    PermissionRequest,
};

// list of op declarations
pub fn ops(super_user: bool) -> Vec<OpDecl> {
    if super_user {
        vec![
            op_check_for_update(),
            op_motd(),
            op_get_current_login(),
            op_get_previous_login(),
            op_login_previous(),
            op_login_new_code(),
            op_login_new_success(),
            op_login_cancel(),
            op_login_guest(),
            op_logout(),
            op_settings(),
            op_set_setting(),
            op_kernel_fetch_headers(),
            op_set_avatar(),
            op_native_input(),
            op_get_bindings(),
            op_set_bindings(),
            op_console_command(),
            op_live_scene_info(),
            op_get_home_scene(),
            op_set_home_scene(),
            op_get_system_action_stream(),
            op_read_system_action_stream(),
            op_get_chat_stream(),
            op_read_chat_stream(),
            op_send_chat(),
            op_get_profile_extras(),
            op_quit(),
            op_get_permission_request_stream(),
            op_read_permission_request_stream(),
            op_get_permission_used_stream(),
            op_read_permission_used_stream(),
            op_set_single_permission(),
            op_set_permanent_permission(),
            op_get_permanent_permissions(),
            op_get_permission_types(),
        ]
    } else {
        Vec::default()
    }
}

#[op2(async)]
#[serde]
async fn op_check_for_update(state: Rc<RefCell<OpState>>) -> Result<(String, String), AnyError> {
    dcl::js::system_api::op_check_for_update(state).await
}

#[op2(async)]
#[string]
async fn op_motd(state: Rc<RefCell<OpState>>) -> Result<String, AnyError> {
    dcl::js::system_api::op_motd(state).await
}

#[op2]
#[string]
fn op_get_current_login(state: &mut OpState) -> Option<String> {
    dcl::js::system_api::op_get_current_login(state)
}

#[op2(async)]
#[string]
async fn op_get_previous_login(state: Rc<RefCell<OpState>>) -> Result<Option<String>, AnyError> {
    dcl::js::system_api::op_get_previous_login(state).await
}

#[op2(async)]
#[serde]
async fn op_login_previous(state: Rc<RefCell<OpState>>) -> Result<(), AnyError> {
    dcl::js::system_api::op_login_previous(state).await
}

#[op2(async)]
#[string]
async fn op_login_new_code(state: Rc<RefCell<OpState>>) -> Result<Option<String>, AnyError> {
    dcl::js::system_api::op_login_new_code(state).await
}

#[op2(async)]
#[string]
async fn op_login_new_success(state: Rc<RefCell<OpState>>) -> Result<(), AnyError> {
    dcl::js::system_api::op_login_new_success(state).await
}

#[op2(fast)]
fn op_login_guest(state: &mut OpState) {
    dcl::js::system_api::op_login_guest(state)
}

#[op2(fast)]
fn op_login_cancel(state: &mut OpState) {
    dcl::js::system_api::op_login_cancel(state)
}

#[op2(fast)]
fn op_logout(state: &mut OpState) {
    dcl::js::system_api::op_logout(state)
}

#[op2(async)]
#[serde]
async fn op_settings(state: Rc<RefCell<OpState>>) -> Result<Vec<SettingInfo>, AnyError> {
    dcl::js::system_api::op_settings(state).await
}

#[op2(async)]
#[serde]
async fn op_set_setting(
    state: Rc<RefCell<OpState>>,
    #[string] name: String,
    val: f32,
) -> Result<(), AnyError> {
    dcl::js::system_api::op_set_setting(state, name, val).await
}

#[op2(async)]
#[serde]
pub async fn op_kernel_fetch_headers(
    state: Rc<RefCell<OpState>>,
    #[string] uri: String,
    #[string] method: Option<String>,
    #[string] meta: Option<String>,
) -> Result<Vec<(String, String)>, AnyError> {
    dcl::js::system_api::op_kernel_fetch_headers(state, uri, method, meta).await
}

#[op2(async)]
pub async fn op_set_avatar(
    state: Rc<RefCell<OpState>>,
    #[serde] base: Option<PbAvatarBase>,
    #[serde] equip: Option<PbAvatarEquippedData>,
    has_claimed_name: Option<bool>,
    #[serde] profile_extras: Option<std::collections::HashMap<String, serde_json::Value>>,
) -> Result<u32, anyhow::Error> {
    dcl::js::system_api::op_set_avatar(state, base, equip, has_claimed_name, profile_extras).await
}

#[op2(async)]
#[string]
pub async fn op_native_input(state: Rc<RefCell<OpState>>) -> String {
    dcl::js::system_api::op_native_input(state).await
}

#[op2(async)]
#[serde]
pub async fn op_get_bindings(state: Rc<RefCell<OpState>>) -> Result<JsBindingsData, anyhow::Error> {
    dcl::js::system_api::op_get_bindings(state).await
}

#[op2(async)]
#[serde]
pub async fn op_set_bindings(
    state: Rc<RefCell<OpState>>,
    #[serde] bindings: JsBindingsData,
) -> Result<(), anyhow::Error> {
    dcl::js::system_api::op_set_bindings(state, bindings).await
}

#[op2(async)]
#[string]
pub async fn op_console_command(
    state: Rc<RefCell<OpState>>,
    #[string] cmd: String,
    #[serde] args: Vec<String>,
) -> Result<String, anyhow::Error> {
    dcl::js::system_api::op_console_command(state, cmd, args).await
}

#[op2(async)]
#[serde]
pub async fn op_live_scene_info(
    state: Rc<RefCell<OpState>>,
) -> Result<Vec<LiveSceneInfo>, anyhow::Error> {
    dcl::js::system_api::op_live_scene_info(state).await
}

#[op2(async)]
#[serde]
pub async fn op_get_home_scene(state: Rc<RefCell<OpState>>) -> Result<HomeScene, anyhow::Error> {
    dcl::js::system_api::op_get_home_scene(state).await
}

#[op2]
pub fn op_set_home_scene(
    state: Rc<RefCell<OpState>>,
    #[string] realm: String,
    #[serde] parcel: Vector2,
) {
    dcl::js::system_api::op_set_home_scene(state, realm, parcel);
}

#[op2(async)]
pub async fn op_get_system_action_stream(state: Rc<RefCell<OpState>>) -> u32 {
    dcl::js::system_api::op_get_system_action_stream(state).await
}

#[op2(async)]
#[serde]
pub async fn op_read_system_action_stream(
    state: Rc<RefCell<OpState>>,
    rid: u32,
) -> Result<Option<SystemActionEvent>, deno_core::anyhow::Error> {
    dcl::js::system_api::op_read_system_action_stream(state, rid).await
}

#[op2(async)]
pub async fn op_get_chat_stream(state: Rc<RefCell<OpState>>) -> u32 {
    dcl::js::system_api::op_get_chat_stream(state).await
}

#[op2(async)]
#[serde]
pub async fn op_read_chat_stream(
    state: Rc<RefCell<OpState>>,
    rid: u32,
) -> Result<Option<ChatMessage>, deno_core::anyhow::Error> {
    dcl::js::system_api::op_read_chat_stream(state, rid).await
}

#[op2(fast)]
pub fn op_send_chat(
    state: Rc<RefCell<OpState>>,
    #[string] message: String,
    #[string] channel: String,
) {
    dcl::js::system_api::op_send_chat(state, message, channel)
}

#[op2(async)]
#[serde]
pub async fn op_get_profile_extras(
    state: Rc<RefCell<OpState>>,
) -> Result<std::collections::HashMap<String, serde_json::Value>, deno_core::anyhow::Error> {
    dcl::js::system_api::op_get_profile_extras(state).await
}

#[op2(fast)]
pub fn op_quit(state: Rc<RefCell<OpState>>) {
    dcl::js::system_api::op_quit(state);
}

#[op2(async)]
pub async fn op_get_permission_request_stream(state: Rc<RefCell<OpState>>) -> u32 {
    dcl::js::system_api::op_get_permission_request_stream(state).await
}

#[op2(async)]
#[serde]
pub async fn op_read_permission_request_stream(
    state: Rc<RefCell<OpState>>,
    rid: u32,
) -> Result<Option<PermissionRequest>, deno_core::anyhow::Error> {
    dcl::js::system_api::op_read_permission_request_stream(state, rid).await
}

#[op2(async)]
pub async fn op_get_permission_used_stream(state: Rc<RefCell<OpState>>) -> u32 {
    dcl::js::system_api::op_get_permission_used_stream(state).await
}

#[op2(async)]
#[serde]
pub async fn op_read_permission_used_stream(
    state: Rc<RefCell<OpState>>,
    rid: u32,
) -> Result<Option<PermissionUsed>, deno_core::anyhow::Error> {
    dcl::js::system_api::op_read_permission_used_stream(state, rid).await
}

#[op2(fast)]
pub fn op_set_single_permission(state: Rc<RefCell<OpState>>, #[bigint] id: usize, allow: bool) {
    dcl::js::system_api::op_set_single_permission(state, id, allow);
}

#[op2]
#[serde]
pub fn op_set_permanent_permission(
    state: Rc<RefCell<OpState>>,
    #[string] level: &str,
    #[string] value: Option<String>,
    #[serde] permission_type: PermissionType,
    #[serde] allow: Option<PermissionValue>,
) -> Result<(), anyhow::Error> {
    dcl::js::system_api::op_set_permanent_permission(state, level, value, permission_type, allow)
}

#[op2(async)]
#[serde]
pub async fn op_get_permanent_permissions(
    state: Rc<RefCell<OpState>>,
    #[string] level: String,
    #[string] value: Option<String>,
) -> Result<Vec<PermanentPermissionItem>, anyhow::Error> {
    dcl::js::system_api::op_get_permanent_permissions(state, &level, value).await
}

#[op2]
#[serde]
pub fn op_get_permission_types() -> Vec<PermissionTypeDetail> {
    dcl::js::system_api::op_get_permission_types()
}
